lib/shared/pe_status_check.rb

```diff
@@ -66,33 +71,43 @@ def self.service_file_exist?(configfile)
     File.exist?("#{configdir}/#{configfile}")
   end
 
-  # Queries the passed port and endpoint for the status API
+  # Module method to make a GET request to an api specified by path and port params
   # @return [Hash] Response body of the API call
-  # param port [Integer] The status API port to query
-  # @param endpoint [String] The status API endpoint to query
-  def self.status_check(port, endpoint)
-    require 'json'
-
-    host     = Puppet[:certname]
-    client   = Puppet.runtime[:http]
-    response = client.get(URI(Puppet::Util.uri_encode("https://#{host}:#{port}/status/v1/services/#{endpoint}")))
-    status   = JSON.parse(response.body)
-    status
-  rescue Puppet::HTTP::ResponseError => e
-    Facter.debug("fact 'self_service' - HTTP: #{e.response.code} #{e.response.reason}")
-  rescue Puppet::HTTP::ConnectionError => e
-    Facter.debug("fact 'self_service' - Connection error: #{e.message}")
-  rescue Puppet::SSL::SSLError => e
-    Facter.debug("fact 'self_service' - SSL error: #{e.message}")
-  rescue Puppet::HTTP::HTTPError => e
-    Facter.debug("fact 'self_service' - General HTTP error: #{e.message}")
-  rescue JSON::ParserError => e
-    Facter.debug("fact 'self_service' - Could not parse body for JSON: #{e}")
+  # @param path [String] The API path to query.  Should include a '/' prefix and query parameters
+  # @param port [Integer] The port to use
+  # @param host [String] The FQDN to use in making the connection.  Defaults to the Puppet certname
+  def http_get(path, port, host = Puppet[:certname])
+    # Use an instance variable to only create an SSLContext once
+    @ssl_context ||= Puppet::SSL::SSLContext.new(
+      cacerts: Puppet[:localcacert],
+      private_key: OpenSSL::PKey::RSA.new(File.read(Puppet[:hostprivkey])),
+      client_cert: OpenSSL::X509::Certificate.new(File.open(Puppet[:hostcert])),
+    )
+
+    client = Net::HTTP.new(host, port)
+    # The main reason to use this approach is to set open and read timeouts to a small value
+    # Puppet's HTTP client does not allow access to these
+    client.open_timeout = 2
+    client.read_timeout = 2
+    client.use_ssl = true
+    client.verify_mode = OpenSSL::SSL::VERIFY_PEER
+    client.cert = @ssl_context.client_cert
+    client.key = @ssl_context.private_key
+    client.ca_file = @ssl_context.cacerts
+
+    response = client.request_get(Puppet::Util.uri_encode(path))
+    if response.is_a? Net::HTTPSuccess
+      JSON.parse(response.body)
+    else
+      false
+    end
+  rescue StandardError => e
+    Facter.warn("Error in fact 'self_service': #{e.backtrace}")
```

Comment:
`self_service` needs to be replaced with `pe_status_check`.

We should stick the backtrace in `debug` level and put a short readable message in `warn` that indicates what it was querying.

Reply:
I swear I put `e.message` in the warn, but good catch.